pygsti/forwardsims/torchfwdsim.py

```diff
@@ -19,206 +25,240 @@
     from pygsti.layouts.copalayout import CircuitOutcomeProbabilityArrayLayout
     import torch
 
-import numpy as np
 import warnings as warnings
-from collections import OrderedDict
 from pygsti.modelmembers.torchable import Torchable
 from pygsti.forwardsims.forwardsim import ForwardSimulator
 
+try:
+    import torch
+    TORCH_ENABLED = True
+except ImportError:
+    TORCH_ENABLED = False
+    pass
+
 
 class StatelessCircuit:
     """
-    Helper data structure useful for simulating a specific circuit quantum (including prep,
+    Helper data structure useful for simulating a specific quantum circuit (including prep,
     applying a sequence of gates, and applying a POVM to the output of the last gate).
     
     The forward simulation can only be done when we have access to a dict that maps
-    pyGSTi Labels to certain PyTorch Tensors.
+    pyGSTi Labels to certain PyTorch Tensors. Construction of that dict is managed by the
+    StatelessModel class, defined below.
     """
 
     def __init__(self, spc: SeparatePOVMCircuit):
         self.prep_label = spc.circuit_without_povm[0]
-        self.op_labels  = spc.circuit_without_povm[1:]
+        if len(spc.circuit_without_povm) > 1:
+            self.op_labels  = spc.circuit_without_povm[1:]
+        else:
+            # Importing this at the top of the file would create a circular dependency.
+            from pygsti.circuits.circuit import Circuit
+            self.op_labels = Circuit(tuple())
         self.povm_label = spc.povm_label
+        self.outcome_probs_dim = len(spc.effect_labels)
         return
+    
+    def outcome_probs(self, torch_bases: Dict[Label, torch.Tensor]) ->  torch.Tensor:
```

Comment:
I am not a fan of this change, and am coincidentally presently advocating for removing a comparable method from the `Circuit` class as part of PR #445. TLDR: In the abstraction hierarchy we typically use nowadays `Circuits` are used for description of a quantum process in terms of elementary operations. Models on the other hand, among other things, are responsible for taking as input that description of a quantum process and returning a probability distribution (part of which it itself farms out to a child ForwardSimulator).

Reply:
Understood. I've just pushed a commit that goes back to the old approach. Everything now happens directly in ``StatelessModel.circuit_probs_from_torch_bases``.